### PR TITLE
Multiupload utility function

### PR DESCRIPTION
**Pull request by rpalaniappan**
- Multi upload utility function

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @rpalaniappan has signed the CLA. Thanks for the pull request!

File: Box.V2.Test.Integration/BoxFilesManagerTestIntegration.cs [hunk 1]
```diff
@@ -254,10 +255,10 @@ public class BoxFilesManagerTestIntegration : BoxResourceManagerTestIntegration
             string partSize = boxFileUploadSession.PartSize;
             long partSizeLong;
             long.TryParse(partSize, out partSizeLong);
-            int numberOfParts = GetNumberOfPartsToBeUploaded(fileSize, partSizeLong);
+            int numberOfParts = Helper.GetNumberOfParts(fileSize, partSizeLong);
 
             // Upload parts in the session
-            await UploadPartsInSession(uploadPartUri, numberOfParts, partSizeLong, fileInMemoryStream, fileSize);
+            await UploadPartsInSessionAsync(uploadPartUri, numberOfParts, partSizeLong, fileInMemoryStream, fileSize);
```
Comment (coolboy) [inline Box.V2.Test.Integration/BoxFilesManagerTestIntegration.cs]: Should we check the response?

File: Box.V2.Test.Integration/BoxFilesManagerTestIntegration.cs [hunk 2]
```diff
@@ -345,15 +346,15 @@ private MemoryStream GetBigFileInMemoryStream(long fileSize)
...
         }
 
-        private async Task UploadPartsInSession(Uri uploadPartsUri, int numberOfParts, long partSize, Stream stream,
+        private async Task UploadPartsInSessionAsync(Uri uploadPartsUri, int numberOfParts, long partSize, Stream stream,
             long fileSize)
         {
             for (int i = 0; i < numberOfParts; i++)
             {
-                string uniqueRandomPartId = GetRandomString(8);
+                string uniqueRandomPartId = Helper.GetRandomString(8);
```
Comment (coolboy) [inline Box.V2.Test.Integration/BoxFilesManagerTestIntegration.cs]: What's the status of partid? are we going to deprecate it?

File: Box.V2/Managers/BoxFilesManager.cs [hunk 3]
```diff
@@ -367,6 +367,64 @@ public BoxFilesManager(IBoxConfig config, IBoxService service, IBoxConverter con
             return response.ResponseObject;
         }
 
+        /// <summary>
+        /// Upload a large file by splitting them up and uploads in a session. NOTE: Available only in .NET CORE. Not available in portable library.
```
Comment (coolboy) [inline Box.V2/Managers/BoxFilesManager.cs]: The function should be available to all our supported platform

File: Box.V2/Managers/BoxFilesManager.cs [hunk 4]
```diff
@@ -367,6 +367,64 @@ public BoxFilesManager(IBoxConfig config, IBoxService service, IBoxConverter con
...
+        /// </summary>
+        /// <param name="stream">The file stream.</param>
+        /// <param name="fileName">Name of the remote file name.</param>
+        /// <param name="fileSize">Size of the file.</param>
+        /// <param name="folderId">parent folder id.</param>
+        /// <returns></returns>
```
Comment (coolboy) [inline Box.V2/Managers/BoxFilesManager.cs]: missing the comment, and it should return the BoxFile

File: Box.V2/Managers/BoxFilesManager.cs [hunk 5]
```diff
@@ -367,6 +367,64 @@ public BoxFilesManager(IBoxConfig config, IBoxService service, IBoxConverter con
...
+        public async Task<bool> UploadUsingSessionAsync(Stream stream, string fileName, long fileSize, string folderId)
+        {
+            // Create Upload Session
+            BoxFileUploadSessionRequest uploadSessionRequest = new BoxFileUploadSessionRequest() {FileName = fileName, FileSize = fileSize, FolderId = folderId};
```
Comment (coolboy) [inline Box.V2/Managers/BoxFilesManager.cs]: Just my curiosity, why not using var?

File: Box.V2/Managers/BoxFilesManager.cs [hunk 6]
```diff
@@ -367,6 +367,64 @@ public BoxFilesManager(IBoxConfig config, IBoxService service, IBoxConverter con
...
+            BoxFileUploadSession boxFileUploadSession = await CreateUploadSessionAsync(uploadSessionRequest);
+
+            // Parse upload session response
+            BoxSessionEndpoint boxSessionEndpoint = boxFileUploadSession.SessionEndpoints;
+            Uri listPartsUri = new Uri(boxSessionEndpoint.ListParts);
+            Uri uploadPartUri = new Uri(boxSessionEndpoint.UploadPart);
+            Uri commitUri = new Uri(boxSessionEndpoint.Commit);
+            string partSize = boxFileUploadSession.PartSize;
+            long partSizeLong;
+            long.TryParse(partSize, out partSizeLong);
```
Comment (coolboy) [inline Box.V2/Managers/BoxFilesManager.cs]: should we check the result of tryParse?

File: Box.V2/Managers/BoxFilesManager.cs [hunk 7]
```diff
@@ -367,6 +367,64 @@ public BoxFilesManager(IBoxConfig config, IBoxService service, IBoxConverter con
...
+            int numberOfParts = Helper.GetNumberOfParts(fileSize, partSizeLong);
+
+            // Upload parts in the session
+            await UploadPartsInSessionAsync(uploadPartUri, numberOfParts, partSizeLong, stream, fileSize);
+
+            // Get the list of parts uploaded in Session
+            // Get upload parts by multiples of 1000 as 1000 is the default
+            List<BoxSessionPartInfo> allSessionParts = new List<BoxSessionPartInfo>();
+            BoxSessionParts boxSessionParts = await GetSessionUploadedPartsAsync(listPartsUri);
```
Comment (coolboy) [inline Box.V2/Managers/BoxFilesManager.cs]: we have auto-paging support in our net sdk, maybe we could improve the code here, let's talk about this

File: Box.V2/Utility/CrossPlatform.cs [hunk 8]
```diff
@@ -34,5 +36,24 @@ public static bool CanConvert<T>(Type objectType)
             return typeof(T).IsAssignableFrom(objectType);
 #endif
         }
+
+        /// <summary>
+        /// Returns the message digest of the stream, formatted as specified by RFC 3230.
+        /// </summary>
+        /// <param name="stream"></param>
+        /// <returns></returns>
```
Comment (coolboy) [inline Box.V2/Utility/CrossPlatform.cs]: looks like you removed it and make it into public interface, let's talk about this

File: Box.V2/Utility/Helper.cs [hunk 9]
```diff
@@ -48,5 +50,59 @@ public static string Base64Encode(string plainText)
                     pair => Uri.UnescapeDataString(pair[ 1 ]));
         }
 
+        /// <summary>
+        /// Retuns the number of parts to be uploaded in Session
+        /// </summary>
+        /// <param name="totalSize"></param>
```
Comment (coolboy) [inline Box.V2/Utility/Helper.cs]: update the comments

File: Box.V2/Utility/Helper.cs [hunk 10]
```diff
@@ -48,5 +50,59 @@ public static string Base64Encode(string plainText)
...
+        /// <param name="partSize"></param>
+        /// <returns></returns>
+        /// <exception cref="Exception"></exception>
+        public static int GetNumberOfParts(long totalSize, long partSize)
+        {
+            if (partSize == 0)
```
Comment (coolboy) [inline Box.V2/Utility/Helper.cs]: if partSize <= 0, throw ... ?

File: Box.V2/Utility/Helper.cs [hunk 11]
```diff
@@ -48,5 +50,59 @@ public static string Base64Encode(string plainText)
...
+            string result = String.Concat(buffer.Select(x => x.ToString("X2")).ToArray());
+            if (length % 2 == 0)
+                return result;
+            return result + random.Next(16).ToString("X");
+        }
+
+        /// <summary>
+        /// Returns part of the file starting at offset.
+        /// </summary>
+        /// <param name="stream"></param>
```
Comment (coolboy) [inline Box.V2/Utility/Helper.cs]: Update the comment

File: Box.V2/Utility/Helper.cs [hunk 12]
```diff
@@ -48,5 +50,59 @@ public static string Base64Encode(string plainText)
...
+        /// <param name="partSize"></param>
+        /// <param name="partOffset"></param>
+        /// <returns></returns>
+        public static Stream GetFilePart(Stream stream, long partSize, long partOffset)
```
Comment (coolboy) [inline Box.V2/Utility/Helper.cs]: is there a more efficient way of doing it?

Comment (coolboy) [inline Box.V2/Utility/Helper.cs]: Not sure if I like all these utility functions belong to FilePartUpload as public here, should  be protected/private to FilesManager IMO, let's chat

**Pull request merged**